Azure/sonic-utilities
# BIOS upgrade support in sonic_installer

**- What I did**
- Added BIOS upgrade support in sonic_installer

**- How I did it**
-  "sonic_installer install sonic-broadcom.bin" --> Post installation of the new image, fetch the next_image version and mount sqashfs filesystem.
- With the help of install_component_firmware() pass component name as "BIOS" and mount location.
- Respective vendor API fetch the BIOS image from mounted location and take appropriate decisions (upgrade/Downgrade/Nochange) respectively. 

**- How to verify it**
 - sonic_installer install sonic-broadcom.bin

**- New command output (if the output of a command-line utility has changed)**
```
--------Upgrade BIOS------------------------------------------------------------
   creating: /host/image-master.0-dirty-20190703.033247/platform/x86_64-dell_s6000_s1220-r0/
  inflating: /host/image-master.0-dirty-20190703.033247/platform/x86_64-dell_s6000_s1220-r0/platform-modules-s6000_1.1_amd64.deb
   creating: /host/image-master.0-dirty-20190703.033247/platform/x86_64-accton_as7312_54x-r0/
  inflating: /host/image-master.0-dirty-20190703.033247/platform/x86_64-accton_as7312_54x-r0/sonic-platform-accton-as7312-54x_1.1_amd64.deb
  inflating: /host/image-master.0-dirty-20190703.033247/fs.squashfs
Installed SONiC base image SONiC-OS successfully

Command: grub-set-default --boot-directory=/host 0

Command: rm -rf /host/old_config

Command: cp -ar /etc/sonic /host/old_config

Command: mount /host/image-master.0-dirty-20190703.033247/fs.squashfs /tmp/tmpQEjjBC -t squashfs -o loop


        ********************************************************************
        * Warning - Upgrading BIOS is inherently risky and should only be  *
        * attempted when necessary.  A failure at this upgrade may cause   *
        * a board RMA.  Proceed with caution !                             *
        ********************************************************************

New BIOS image 3.25.0.2-7 available to install, continue? [y/N]: y
Command: /usr/local/bin/flashrom -p internal -w /tmp/tmpQEjjBC/usr/share/sonic/device/x86_64-dell_s6100_c2538-r0/bin/S6100-BIOS-3.25.0.2-7.bin
Erasing and writing flash chip... Erase/write done.
Verifying flash... VERIFIED.

Command: umount /tmp/tmpQEjjBC

Command: sync;sync;sync

Command: sleep 3

Done
```

-->
